modules/basis/basis-datasource/tomcat-listener/src/main/java/io/dirigible/tomcat/listener/NamingResourcesListener.java

```diff
@@ -0,0 +1,156 @@
+import org.apache.catalina.LifecycleListener;
+import org.apache.catalina.deploy.NamingResourcesImpl;
+import org.apache.naming.ContextAccessController;
+import org.apache.tomcat.util.descriptor.web.ContextResource;
+import org.apache.tomcat.util.descriptor.web.ResourceBase;
+
+public class NamingResourcesListener implements LifecycleListener, PropertyChangeListener {
+
+  private static final Logger LOGGER = Logger.getLogger(NamingResourcesListener.class.getName());
+
+  private static final String RESOURCE_STR = "resource";
+
+  private static final String RESOURCE_ENV_REF_STR = "resourceEnvRef";
+
+  static final String FACTORY_CLASS_NAME = "io.dirigible.listener.DelegatingObjectFactory";
+
+  @Deprecated
+  static final String PERSISTENCE_FACTORY_CLASS_NAME = "com.sap.jpaas.service.persistence.core.JNDIDataSourceFactory";
+
+  static final String FACTORY_PROPERTY = "factory";
+
+  private Object container = null;
+
+  private NamingResourcesImpl namingResources = null;
+
+  private boolean initialized = false;
+
+  private static final String name = "/";
+
+  private String factoryClassName;
+
+  static final String RESOURCE_REF_NAME = "ResourceRefName";
+
+  private static final String DATA_SOURCE = "DataSource";
+
+  @Override
+  public void propertyChange(PropertyChangeEvent event) {
+    if (!this.initialized) {
+      return;
+    }
+
+    final Object source = event.getSource();
+    if (source == this.namingResources) {
+
+      // Setting the context in read/write mode
+      ContextAccessController.setWritable(name, this.container);
+
+      final String name = event.getPropertyName();
+      final Object newValue = event.getNewValue();
+
+      if (newValue != null && (RESOURCE_STR.equals(name) || RESOURCE_ENV_REF_STR.equals(name))) {
+        ResourceBase resource = (ResourceBase) newValue;
+        updateResource(resource);
+      }
+
+      // Setting the context in read only mode
+      ContextAccessController.setReadOnly(name);
+
+    }
+  }
+
+  @Override
+  public void lifecycleEvent(LifecycleEvent event) {
+    this.container = event.getLifecycle();
+
+    if (!(this.container instanceof Context)) {
+      return;
+    }
+
+    this.namingResources = ((Context) this.container).getNamingResources();
+
+    if (Lifecycle.CONFIGURE_START_EVENT.equals(event.getType())) {
+      if (this.initialized) {
+        return;
+      }
+
+      this.namingResources.addPropertyChangeListener(this);
+
+      // handle resource-ref
+
+      ResourceBase[] resources = this.namingResources.findResources();
+      updateResources(resources);
+
+      // handle resource-env-ref
+      resources = this.namingResources.findResourceEnvRefs();
+      updateResources(resources);
+
+      this.initialized = true;
+    } else if (Lifecycle.CONFIGURE_STOP_EVENT.equals(event.getType())) {
+      if (!this.initialized) {
+        return;
+      }
+
+      this.namingResources.removePropertyChangeListener(this);
+
+      this.initialized = false;
+    }
```

Comment:
Can it be refactored without multiple return statements?

Reply:
done